### PR TITLE
Added automatic token refreshing if token is expired

### PR DESCRIPTION
**Pull request by oguzcankirmemis**
Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

### CODE REVIEW

**Comment by oguzcankirmemis** [comment]
@wagmarcel @srware Marked as draft, until the [PR](https://github.com/Open-IoT-Service-Platform/oisp-sdk-js/pull/36) is merged. It is mandatory for this update to work.

File: lib/cloud.proxy.js [hunk 1]
```diff
@@ -83,6 +88,47 @@ IoTKitCloud.prototype.isActivated = function () {
     }
     return false;
 };
+
+IoTKitCloud.prototype.ensureFreshness = function (callback) {
```
Comment (srware) [inline lib/cloud.proxy.js:92]: Can you change the name of this function to something more specific like "checkDeviceToken".

"ensureFreshness" looks and sounds weird to me although I understand where you were coming from :)

File: lib/cloud.proxy.js [hunk 2]
```diff
@@ -159,125 +209,152 @@ IoTKitCloud.prototype.setDeviceCredentials = function() {
...
+        me.proxy.attributes(doc, function () {
+            me.logger.debug("Attributes returned from " + me.proxy.type);
+            if (callback) {
+                callback();
+            }
+        });
+    };
+
+    me.ensureFreshness(handler);
 };
```
Comment (srware) [inline lib/cloud.proxy.js:262]: Is it necessary to change this function given it is only ever called from the the update() function and we are checking the device token there first?

File: lib/cloud.proxy.js [hunk 3]
```diff
@@ -159,125 +209,152 @@ IoTKitCloud.prototype.setDeviceCredentials = function() {
...
 
 
 IoTKitCloud.prototype.updateOld = function(callback) {
```
Comment (srware) [inline lib/cloud.proxy.js:265]: Is this function even required anymore...?

**Comment by oguzcankirmemis** [comment]
@srware Also fixed the bugs, the tests do pass without the SDK release, because in our case right now token never expires (10 years lifetime). Nevertheless I tested with the newest SDK and the refreshing mechanism works. In future we should add e2e-test for it too, when we reduce the lifetime of token. Ready for merging.

**Pull request merged**